### DISCUSSION TITLE
Slice best practice?

### DISCUSSION THREAD

**Discussion by hexwit** [pmndrs/zustand#1131] [Q&A]
Hello! 

I am pretty new to stores and such kind of state management. Have no exp with redux. 

So my question is, how it is better to form slices? 
Like one object as 
```
    const state = store(state => {
        return {
            pages: state.pages,
            pageOrder: state.pageOrder,
            activePage: state.activePage
        }
    });
```
or that should be three different slices? 
```
    const pages = store(state => state.pages);
    const pageOrder = store(state => state.pageOrder);
    const activePage = store(state => state.activePage);
```

Is there any different between this approaches? Performance, usability, etc. 
May be some better approach?

**Comment by dai-shi** [comment]
I'd recommend the second approach. The first one causes re-renders every time, because the function returns a new object. (You need to pass `shallow` to the second argument.) So, the second approach is less error-prone.

I also recommend `use` prefix like `useStore(...)` instead of `store(...)`.

FYI: Redux Style Guide also recommends multiple hooks. https://redux.js.org/style-guide/#call-useselector-multiple-times-in-function-components

**Reply by aksalsf** [reply]
Hi @dai-shi . I define my store using slice pattern like below.

```
import create from "zustand"
import { devtools } from "zustand/middleware"

import { TKlasterSlice, createKlasterSlice } from "./klaster-store"
import { TRootSlice, createRootSlice } from "./root-store"

export const useStore = create<TRootSlice & TKlasterSlice>()(
  devtools(
    (...args) => ({
      ...createRootSlice(...args),
      ...createKlasterSlice(...args)
    })
    , { serialize: true })
)

export const {
  getState: getStore,
  setState: updateStore,
  subscribe: subscribeStore,
  destroy: destroyStore
} = useStore

export function cleanupStore () {
  const { clearRootState, clearKlasterState } = getStore()
  destroyStore()
  clearRootState()
  clearKlasterState()
}
```

Each slice is different entity. Like root is for global and klaster is for klaster entity. Is my approach is correct because when I call the store is look mixed because all of the entity store is on one place?

**Reply by dai-shi** [reply]
If root and klaster are totally isolated, you can define two stores. If you want to use `devtools` just for one, the combined store (slices) should be good.

btw, I never used `destroy` method. Usually, correct use of hooks would just be fine.
